test/resolver_test.go

```diff
@@ -0,0 +1,194 @@
+ * limitations under the License.
+ *
+ */
+
+package test
+
+import (
+	"context"
+	"reflect"
+	"testing"
+	"time"
+
+	iresolver "google.golang.org/grpc/internal/resolver"
+	"google.golang.org/grpc/internal/serviceconfig"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/resolver/manual"
+	testpb "google.golang.org/grpc/test/grpc_testing"
+)
+
+type funcConfigSelector struct {
+	f func(iresolver.RPCInfo) *iresolver.RPCConfig
+}
+
+func (f funcConfigSelector) SelectConfig(i iresolver.RPCInfo) *iresolver.RPCConfig {
+	return f.f(i)
+}
+
+func (s) TestConfigSelector(t *testing.T) {
+	gotInfoChan := make(chan *iresolver.RPCInfo, 1)
+	sendConfigChan := make(chan *iresolver.RPCConfig, 1)
+	gotContextChan := make(chan context.Context, 1)
+
+	ss := &stubServer{
+		emptyCall: func(ctx context.Context, in *testpb.Empty) (*testpb.Empty, error) {
+			gotContextChan <- ctx
+			return &testpb.Empty{}, nil
+		},
+	}
+	ss.r = manual.NewBuilderWithScheme("confSel")
+
+	if err := ss.Start(nil); err != nil {
+		t.Fatalf("Error starting endpoint server: %v", err)
+	}
+	defer ss.Stop()
+
+	state := iresolver.SetConfigSelector(resolver.State{
+		Addresses:     []resolver.Address{{Addr: ss.address}},
+		ServiceConfig: parseCfg(ss.r, "{}"),
+	}, funcConfigSelector{
+		f: func(i iresolver.RPCInfo) *iresolver.RPCConfig {
+			gotInfoChan <- &i
+			cfg := <-sendConfigChan
+			if cfg != nil && cfg.Context == nil {
+				cfg.Context = i.Context
+			}
+			return cfg
+		},
+	})
+	ss.r.UpdateState(state) // Blocks until config selector is applied
+
+	longdeadlineCtx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	shorterTimeout := 3 * time.Second
+
+	testMD := metadata.MD{"footest": []string{"bazbar"}}
+	mdOut := metadata.MD{"handler": []string{"value"}}
+
+	var onCommittedCalled bool
+
+	testCases := []struct {
+		name   string
+		md     metadata.MD
+		config *iresolver.RPCConfig
+	}{{
+		name:   "basic",
+		md:     testMD,
+		config: &iresolver.RPCConfig{},
+	}, {
+		name: "alter MD",
+		md:   testMD,
+		config: &iresolver.RPCConfig{
+			Context: metadata.NewOutgoingContext(ctx, mdOut),
+		},
+	}, {
+		name: "alter timeout; remove MD",
+		md:   testMD,
+		config: &iresolver.RPCConfig{
+			Context: longdeadlineCtx,
+		},
+	}, {
+		name:   "nil config",
+		md:     metadata.MD{},
+		config: nil,
+	}, {
+		name: "alter timeout via method config; remove MD",
+		md:   testMD,
+		config: &iresolver.RPCConfig{
+			MethodConfig: serviceconfig.MethodConfig{
+				Timeout: &shorterTimeout,
+			},
+		},
+	}, {
+		name: "onCommitted callback",
+		md:   testMD,
+		config: &iresolver.RPCConfig{
+			OnCommitted: func() {
+				onCommittedCalled = true
+			},
+		},
+	}}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			select {
+			case sendConfigChan <- tc.config:
```

Comment:
This seems to add an unnecessary coupling between test `N` and test `N-1`. Could this be avoided by making sure the expected state is reached at the end of test `N-1`?

Reply:
> Another option would be to recreate the client or update the config selector for every test case - then we wouldn't need a channel. WDYT?

Yes, I think that would be better.

Reply:
Done.  Getting rid of the channels is nicer.